### PR TITLE
Add a flush timeout to config and use it instead of hardcoded 30s

### PR DESCRIPTION
**Pull request by wozzo**
*Issue #180*

*Description of changes:*
Add a configurable flush timeout, replacing hard coded 30s

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

Was unable to get many of the tests passing locally without any code changes.

### CODE REVIEW

File: src/AWS.Logger.AspNetCore/AWSLoggerConfigSection.cs [hunk 1]
```diff
@@ -128,6 +128,7 @@ public class AWSLoggerConfigSection
         internal const string LOG_STREAM_NAME_PREFIX = "LogStreamNamePrefix";
         internal const string LIBRARY_LOG_FILE_NAME = "LibraryLogFileName";
         internal const string LIBRARY_LOG_ERRORS = "LibraryLogErrors";
+        internal const string FLUSH_TIMEOUT = "FlushTimeout";
```
Comment (snakefoot) [inline src/AWS.Logger.AspNetCore/AWSLoggerConfigSection.cs:131]: Maybe call it `FlushTimeoutSeconds` ?

File: src/NLog.AWS.Logger/AWSTarget.cs [hunk 2]
```diff
@@ -200,6 +200,19 @@ public string LibraryLogFileName
...
         }
 
+        /// <summary>
+        /// Gets the FlushTimeout property. When performing a flush of the in-memory queue this is the maximum period of time allowed to send the remaining
+        /// messages before it will be aborted. If this is exceeded, incoming log messages will be dropped.
+        /// <para>
+        /// The default is 30 seconds.
+        /// </para>
+        /// </summary>
+        public TimeSpan FlushTimeout
```
Comment (snakefoot) [inline src/NLog.AWS.Logger/AWSTarget.cs:210]: Maybe call it `FlushTimeoutSeconds` ? (Right now If using FlushTimeout = 30, then it means 30 days).

Comment (wozzo) [reply src/NLog.AWS.Logger/AWSTarget.cs:210]: But it's a timespan, not an int.

Comment (snakefoot) [reply src/NLog.AWS.Logger/AWSTarget.cs:210]: The property-type should ofcourse match the name. Making it easier to configure from `NLog.config`. Like this:
```c#
        public int FlushTimeoutSeconds
        {
            get { return (int)_config.FlushTimeout.TotalSeconds; }
            set { _config.FlushTimeout = TimeSpan.FromSeconds(value); }
        }
```

Comment (wozzo) [reply src/NLog.AWS.Logger/AWSTarget.cs:210]: There are other timespans in the config. I was keeping with the already established approach for configuring timespans for consistency.

Comment (snakefoot) [reply src/NLog.AWS.Logger/AWSTarget.cs:210]: Oh well. Just trying to make things better.

Comment (wozzo) [reply src/NLog.AWS.Logger/AWSTarget.cs:210]: Cheers. Need to get someone from Amazon to look at it, otherwise it'll never get merged either way.

**Comment by ashishdhingra** [comment]
Duplicate of https://github.com/aws/aws-logging-dotnet/pull/174. Both PR(s) need to be reviewed and best changes merged.

**Comment by ashishdhingra** [comment]
@wozzo Please change the target branch to `dev` (instead of `master`) for this PR.

**Comment by wozzo** [comment]
@ashishdhingra done

**Comment by ashishdhingra** [comment]
@wozzo Please have a look at the review comments from @attilah.

File: src/AWS.Logger.AspNetCore/AWSLoggerConfigSection.cs (hunk 1 above)
Comment (ashishdhingra) [reply src/AWS.Logger.AspNetCore/AWSLoggerConfigSection.cs:131]: Agreed

**Comment by ashishdhingra** [comment]
@attilah I have created a new PR https://github.com/aws/aws-logging-dotnet/pull/193 to address your review comments. It retains @wozzo as author. Please review.

File: src/AWS.Logger.Log4net/AWSAppender.cs [hunk 3]
```diff
@@ -223,7 +236,8 @@ public override void ActivateOptions()
 				LogStreamNameSuffix = LogStreamNameSuffix,
                 LogStreamNamePrefix = LogStreamNamePrefix,
                 LibraryLogErrors = LibraryLogErrors,
-				LibraryLogFileName = LibraryLogFileName
+				LibraryLogFileName = LibraryLogFileName,
+                FlushTimeout = FlushTimeout
```
[the remark this answers is not in the document]
Comment (ashishdhingra) [reply src/AWS.Logger.Log4net/AWSAppender.cs:240]: Addressed in PR https://github.com/aws/aws-logging-dotnet/pull/193

File: src/NLog.AWS.Logger/AWSTarget.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (ashishdhingra) [reply src/NLog.AWS.Logger/AWSTarget.cs:210]: Addressed in PR https://github.com/aws/aws-logging-dotnet/pull/193

File: src/AWS.Logger.Core/AWSLoggerConfig.cs [hunk 4]
```diff
@@ -67,9 +67,9 @@ public class AWSLoggerConfig : IAWSLoggerConfig
         /// </summary>
         public string Region { get; set; }
 
-        /// <summary>
-        /// Gets and sets of the ServiceURL property. This is an optional property; change
-        /// it only if you want to try a different service endpoint. Ex. for LocalStack
+        /// <summary>
```
[the remark this answers is not in the document]
Comment (ashishdhingra) [reply src/AWS.Logger.Core/AWSLoggerConfig.cs:70]: Addressed in PR https://github.com/aws/aws-logging-dotnet/pull/193

File: src/AWS.Logger.Core/AWSLoggerConfig.cs [hunk 5]
```diff
@@ -161,19 +161,28 @@ public AWSLoggerConfig(string logGroup)
         public string LogStreamNamePrefix { get; set; } = string.Empty;
 
         /// <summary>
-        /// Gets and sets the LibraryLogErrors property. This is the boolean value of whether or not you would like this library to log logging errors.
+        /// Gets and sets the LibraryLogErrors property. This is the boolean value of whether or not you would like this library to log logging errors.
```
[the remark this answers is not in the document]
Comment (ashishdhingra) [reply src/AWS.Logger.Core/AWSLoggerConfig.cs:164]: Addressed in PR https://github.com/aws/aws-logging-dotnet/pull/193

File: src/AWS.Logger.Core/AWSLoggerConfig.cs [hunk 6]
```diff
@@ -161,19 +161,28 @@ public AWSLoggerConfig(string logGroup)
...
-        /// <summary>
-        /// Gets and sets the LibraryLogFileName property. This is the name of the file into which errors from the AWS.Logger.Core library will be written into.
-        /// <para>
         /// The default is "aws-logger-errors.txt".
         /// </para>
         /// </summary>
         public string LibraryLogFileName { get; set; } = "aws-logger-errors.txt";
+
+        /// <summary>
+        /// Gets the FlushTimeout property. When performing a flush of the in-memory queue this is the maximum period of time allowed to send the remaining
```
[the remark this answers is not in the document]
Comment (ashishdhingra) [reply src/AWS.Logger.Core/AWSLoggerConfig.cs:180]: Addressed in PR https://github.com/aws/aws-logging-dotnet/pull/193

File: src/AWS.Logger.Core/AWSLoggerConfig.cs [hunk 7]
```diff
@@ -161,19 +161,28 @@ public AWSLoggerConfig(string logGroup)
...
+        /// messages before it will be aborted. If this is exceeded, incoming log messages will be dropped.
+        /// <para>
+        /// The default is 30 seconds.
+        /// </para>
+        /// </summary>
+        public TimeSpan FlushTimeout { get; set; } = TimeSpan.FromSeconds(30);
```
[the remark this answers is not in the document]
Comment (ashishdhingra) [reply src/AWS.Logger.Core/AWSLoggerConfig.cs:186]: Addressed in PR https://github.com/aws/aws-logging-dotnet/pull/193

**Comment by wozzo** [comment]
Cheers @ashishdhingra. I hadn't gotten around to this yet.

**Comment by ashishdhingra** [comment]
Closing this in favor of https://github.com/aws/aws-logging-dotnet/pull/193.

**Pull request closed without merging**